R/scale-discrete-.r

```diff
@@ -132,6 +134,14 @@ ScaleDiscretePosition <- ggproto("ScaleDiscretePosition", ScaleDiscrete,
     }
   },
 
+  get_breaks = function(self, limits = self$get_limits()) {
+    if (self$warn && length(self$get_limits()) == 0) {
```

Comment:
If you use `warn`, I think you'd need to reset it in the `clone()` method below.

What if in `Scale$train_df()` we called `train_init()` and `train_final()` methods? (or begin/end, setup/teardown, start/finish) Then you could call this in `train_finalise()`. (I'm not sure what you'd use `train_init()` for, but it makes sense for symmetry)